nixos/dovecot: add settings option (RFC42)

Jappie3 opened the pull request:
follow-up for #388463, this adds an RFC42-style `settings` option to the Dovecot module

until that PR is merged, this one will include all of its changes in the diff (which is every commit except the last one)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Jappie3 committed:
dovecot-fts-flatcurve: only build for dovecot 2.3

Jappie3 committed:
nixos/dovecot: update for 2.4

Jappie3 commented:
apologies for the pings, I forgot to mark this PR as draft when creating it

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -401,18 +283,104 @@ in
+            { name, ... }:
+            let
+              # splits name on the first space
+              # and return null when name is a list - e.g. [definition 1-entry 1]
+              splits = if builtins.isList name then null else builtins.match "([^ ]+) (.+)" name;
```
1. builtins.isList checks whether the argument is an object of type list, this will never be true because "name" is always a string
2. with a slightly modified submodule type, you can extract the intended name even when the container of this is a list, so this handles all cases and you can remove the parsing in formatKeyValue. See https://github.com/2xsaiko/nixpkgs/commit/push-nplnzxkytsqp

Jappie3 replied:
yo I love that custom merge function approach - thanks!

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -828,8 +799,8 @@ in
       }
-      // optionalAttrs (cfg.createMailUser && cfg.mailGroup != null) {
-        ${cfg.mailGroup} = { };
+      // optionalAttrs (cfg.settings ? mail_gid && cfg.settings.mail_gid != null && cfg.createMailUser) {
```
```suggestion
      // optionalAttrs (cfg.settings.mail_gid or null != null && cfg.createMailUser) {
```

A bit less repetitive. (also above for mail_uid and mail_gid)

MattSturgeon commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -401,18 +283,104 @@ in
+            }
+          );
+          valueType =
+            nullOr (oneOf [
```
> Maybe we should disallow values of [] since it's ambiguous

IIRC there's a `types.nonEmptyListOf` or something.

Jappie3 replied:
seems like with the current version of this code, `test = [];` is parsed to `test =`

I don't see why anyone would want an empty list as dovecot config value - if they want to create something like `foo = ` they can use an empty string I think - so using `types.nonEmptyListOf primitiveType` sounds like a good idea

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -106,121 +93,64 @@ let
+      let
+        # Use the first component of the attr name as the section type, and the
+        # rest if present as the section name
+        splits = lib.elemAt (lib.elemAt (lib.split "^([^ ]+)( +(.+))?$" n) 1);
```
That's what this already does, this isn't concatenating anything. As far as I can tell your code is equivalent, I'm not sure what you mean here.

Jappie3 replied:
first of all: i'm going to remove this logic at line 120 & do the regex splitting at line 330, currently it's duped cuz I forgot to remove the upper lines.

> Yes it's equivalent. I just prefer an implementation that doesn't split then concat the name part.

oh wait you mean this line?

```
${if sectionName == null then "" else "${removePrefix " " sectionName} "}
```

@2xsaiko suggested the following somewhere:

```nix
        sectionType = v.section.type;
        sectionName = v.section.name;
        sectionTitle = concatStringsSep " " (
          filter (s: s != null) [
            sectionType
            sectionName
          ]
        );
```

which is way cleaner indeed

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -272,22 +206,215 @@ let
+            { name, ... }:
+            let
+              # splits name on the first space
+              splits = builtins.match "([^ ]+) (.+)" name;
```
```suggestion
              splits = builtins.match "([^ ]+) +(.+)" name;
```

see https://github.com/NixOS/nixpkgs/pull/410912#discussion_r2107669499. On second thought, not sure if we *should* do this since it makes keys `a b` and `a  b` map to the same section in the config.

For the current code to not do the same though in the end, it would have to implement escaping in the config printer though ([function settings_section_escape in settings-parser.c](https://github.com/dovecot/core/blob/42864e4182c7bfe2931fdf888f3cb4ba77953dee/src/lib-settings/settings-parser.c#L1153)). Maybe a good idea to implement this because sometimes the section name is interpreted so you might want some of those special chars in the name.

MattSturgeon replied:
If it uses backslash escapes, we should be able to use `lib.strings.escape`.

E.g. `escape [ "/" "+" "." ] name`

Jappie3 replied:
do you mean something like this?

```
nix-repl> :p builtins.replaceStrings ["=" ''\'' " " ","] [''\e'' ''\\'' ''\_'' ''\+''] ''test \ s.ect_ion , na=me''
test\_\\\_s.ect_ion\_\+\_na\eme
```

MattSturgeon commented:
> quota breaks - options got removed
>
> ssl options break - options got removed

I'd look in a bit more detail once the PR is more finalised, but it should be possible to provide a sane migration path for removed/deprecated options using things like `mkDerivedConfig`, `mkChangedOptionModule`, `mkRenamedOptionModule`, etc.

Jappie3 commented:
quota is probably possible, but I'm totally unfamiliar with that plugin so haven't looked too much into it yet

SSL is more complicated, the options got renamed in 2.4 (see https://doc.dovecot.org/2.4.0/installation/upgrade/2.3-to-2.4.html#converted-settings)

if you want to do something like `mkRenamedOptionModule`, you need to know what to rename the option to and thus you need to evaluate `package.version`, which is - as far as I know - not possible since that creates a circular dependency -> infrec

fwiw, these are the options in question:

2.3:
`cfg.ssl_ca`, `cfg.ssl_cert`, `cfg.ssl_key`

2.4:
`cfg.settings.ssl_server_ca_file`, `cfg.settings.ssl_server_cert_file`, `cfg.settings.ssl_server_key_file`

Jappie3 commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -106,121 +96,67 @@ let
+    indent: n: v:
+    if (v == null) then
+      ""
+    else if isInt v then
+      toOption indent n (toString v)
```
only ints iirc, I've never seen any floats

https://doc.dovecot.org/2.4.0/core/settings/types.html also doesn't list floats

Jappie3 commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -106,121 +96,67 @@ let
+      toOption indent n (toString v)
+    else if isBool v then
+      toOption indent n (yesOrNo v)
+    else if isString v then
+      toOption indent n v
```
I added that derivation check to make sure the parser didn't descend into the attrs of a derivation, but yes - handling paths as well is definitely a good idea

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -272,22 +208,204 @@ let
+            ;
+          inherit (lib.lists) last length sublist;
+
+          # Override submodule to pass the parent name to the submodule name arg
+          # if the current name is a list (starting with '[definition ') so that
+          # it can be parsed by the submodule
+          submodule' =
+            f:
+            let
+              orig = submodule f;
+              merge =
+                loc: defs:
+                let
+                  loc' =
+                    if builtins.match ''\[definition .*'' (last loc) != null then
+                      sublist 0 (length loc - 1) loc
+                    else
+                      loc;
+                in
+                orig.merge loc' defs;
+            in
+            orig // { inherit merge; };
```
> options._module.args.loc

Ohhh what, that does get passed to the submodule, but only like that? I tried adding loc as a module argument directly since that's what _module.args is supposed to do but it didn't work iirc so I did it this way.

But yes, that is the point.

MattSturgeon replied:
> > options._module.args.loc
> 
> Ohhh what, that does get passed to the submodule, but only like that?

It gets passed to the submodule via `lib.evalModules`' `prefix` argument.

IIRC `prefix` isn't made available to modules directly (I could be wrong on that), but it _is_ prefixed to all options' `loc`.

The `_modules.args` option is just an arbitrary option known to always exist, since it is declared by `evalModules` itself.

2xsaiko replied:
Yeah, I think I saw that, and that's why I tried adding it as a module arg, i.e. `{config, loc, ...}: {…}`, because at least in top level NixOS modules everything under _module.args can be accessed like that. Is there something different about this one?

2xsaiko replied:
I'm very familiar with options vs. config and the option metadata. Just didn't know you could put custom values there, and it certainly seems very unconventional to do it like this.

MattSturgeon replied:
I'm not sure what you mean. There are no custom values here. `loc` is always available on all options.

MattSturgeon replied:
`options._module.args.loc` is the location of the `_module.args` option. This _happens_ to include the location of the submodule as its prefix.

The location is `prefix ++ [ "_module" "args" ]`.

All option locations include the `prefix` passed to `evalModules`. There's nothing special about the `_module.args` option; I just chose it because all module evals have that option.

We could also use `options._module.check.lock` or `options.section.name.lock` or any other option in the submodule.

MattSturgeon commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -272,22 +210,195 @@ let
+              # if the current name is a list (starting with '[definition ') -> get
+              # name' from _module.args.loc & use it for {type,name}Default
+              name' =
+                if (builtins.match ''\[definition .*'' name) == null then
```
If you only care about the start, you can use `lib.strings.hasPrefix` (not `lib.lists.hasPrefix`).

```nix
lib.strings.hasPrefix "[definition " name
```

However you could also use a more precies regex instead:

```nix
builtins.match ''\[definition .*\\]'' name
```

MattSturgeon commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -272,22 +210,195 @@ let
+                else
+                  last (dropEnd 3 options._module.args.loc);
+
+              # split name' on the first space
+              splits = builtins.match "([^ ]+) +(.+)" name';
+              typeDefault = if splits == null then name' else builtins.elemAt splits 0;
+              nameDefault = if splits == null then null else builtins.elemAt splits 1;
```
As per https://github.com/NixOS/nixpkgs/pull/410912#discussion_r2107806459, should we drop the ` +` ?

```suggestion
              # split name' on the first space
              splits = builtins.match "([^ ]+) (.+)" name';
              typeDefault = if splits == null then name' else builtins.elemAt splits 0;
              nameDefault = if splits == null then null else builtins.elemAt splits 1;
```

MattSturgeon commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -272,22 +208,204 @@ let
+            in
+            {
+              options = {
+                section = {
```
I think naming it something like `_section` may help make it clear to any users inspecting their config that this is a special attribute, rather than a normal freeform value.

Most of the time users won't be explicitly defining `section` (or `_section`) themselves, right?

Jappie3 committed:
dovecot: add 2.4.0 alongside 2.3.21.1

dblsaiko reviewed (changes requested):
Finally porting my config and trying this out. Sorry for the wait!

I’ve got it to be mostly free of everything outside of settings now and it seems to work fine after figuring out the imapsieve insanity. YMMV.

https://git.dblsaiko.net/systems/commit/?h=push-pmtkvsqnqlzn

I made some changes in the module to get it to work, I think I put all of it in the review comments here though. https://github.com/2xsaiko/nixpkgs/commit/push-nnmrwwtsowql

---

Also, this definitely needs to be looked into: you can’t put bare paths in cfg.settings or it will try to import (and interpret them as a NixOS module, from what I can see in the stack trace) for some reason. What the fuck.

```
settings = {
  blargh = ./dovecot/report-spam.sieve;
}
```

```
       error: syntax error, expecting ']'
       at /nix/store/0jhxh3sg3j00jw0krdvd2602vwalqygk-wsnj7c1kijf94j2i37hdnpmxkxgqwjzv-source/configurations/polaris/dovecot/report-spam.sieve:1:28:
            1| require ["vnd.dovecot.pipe", "copy", "imapsieve", "environment", "variables"];
             |                            ^
            2|
```

Putting the path into "${…}" works.

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -272,22 +211,195 @@ let
+            }
+          );
+
+          primitiveType = oneOf [
+            int
+            str
+            bool
+            section
+          ];
```
Oh what, you can actually use a path in place of a submodule? I thought that might be what was happening but I’ve never seen that actually done anywhere so that was just a theory. I thought it was only possible in imports.

https://gist.github.com/2xsaiko/93b5a91c8e43109a15a865f7e000399d

Jappie3 committed:
dovecot_pigeonhole: patch regarding permission error

This patch fixes a permission issue that occurs when saving compiled
sieve scripts sourced from the nix store. Instead of reusing the
read-only permission bits from the nix store, it explicitly uses `0700`
for the directory in which compiled sieve scripts should be saved.

Additional context:
- https://github.com/NixOS/nixpkgs/pull/388463#issuecomment-3066016707
- https://github.com/dovecot/pigeonhole/pull/15

Jappie3 commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -625,37 +658,138 @@ in
+        # these options differ quite a bit between 2.3 and 2.4, which is why they were split up here
+        if versionAtLeast cfg.package.version "2.4" then
+          {
+            mail_home = mkDefault "/var/spool/mail/%{user}";
```
are you sure about this? the dovecot documentation gives the following example:

```
# https://doc.dovecot.org/2.4.0/core/summaries/settings.html#mail_home

mail_home = /var/vmail/%{user | domain}/%{user | username}
mail_path = ~/mail
```

~~where `mail_home` is the directory where mail is stored, and `mail_path` specifies the per-user subdirectory of `mail_home`~~

ok hold on did they actually switch the 2 options in that example? because [`mail_path` mentions](https://doc.dovecot.org/2.4.0/core/summaries/settings.html#mail_path):

> Path to a directory where the mail is stored. [Mail User Variables](https://doc.dovecot.org/2.4.0/core/settings/variables.html#mail-user-variables) are commonly used here.

> Usually the mails should be stored in a sub-directory under the home directory, but not the home directory itself (see [Home Directories for Virtual Users](https://doc.dovecot.org/2.4.0/core/config/users/virtual.html#home-directories)).

which makes me think that `mail_path` should indeed be the absolute directory and `mail_home` the per-user directory under `mail_path`...

yeaaaah...

> Home directory is a per-user directory where Dovecot can save user-specific files.

source: https://doc.dovecot.org/2.4.0/core/config/users/virtual.html#home-directories

Jappie3 replied:
ok counterargument: the following config:

```
mail_home = "/var/vmail/%{user | domain}/%{user | username}";
mail_path = "~/mail";
```

produces the following directory structure:

```
$> tree /var/vmail/jappie.dev/jasper/
/var/vmail/jappie.dev/jasper/
└── mail
    ├── cur
        [...]
    ├── dovecot.index
    ├── dovecot.index.cache
    ├── dovecot.index.log
    ├── dovecot-keywords
    ├── dovecot.list.index
    ├── dovecot.list.index.log
    ├── dovecot.mailbox.log
    ├── dovecot-uidlist
    ├── dovecot-uidvalidity
    ├── dovecot-uidvalidity.67fa43bb
    ├── maildirfolder
    ├── new
    ├── subscriptions
    └── tmp
```

which seems correct? `<domain>/<user>/mail`

2xsaiko replied:
mail_path is where it stores mail, and the default is /var/spool/mail/…. mail_home is where “~” resolves to in the config, the default is the user’s home directory. This is the old default as well, I think.

I’d have to check again but I’m pretty sure I had to change this back myself so it would work with the old default that I used (I think that was related to sieve scripts since I always specified them relative to ~ and suddenly that was no longer the user home)

Jappie3 commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -625,37 +658,138 @@ in
+        listToAttrs (
+          map (m: {
+            name = "protocol ${elemAt (splitString "." m.name) 0}";
+            value.mail_plugins = "$mail_plugins ${
+              concatStringsSep " " (m.value.enable ++ cfg.mailPlugins.globally.enable)
+            }";
```
I've added an assertion to check for v2.4 & cfg.mailPlugins (which includes both the per-protocol & global options)

Jappie3 commented:
fixed ci (formatting)

Jappie3 committed:
dovecot_exporter: add for dovecot 2.4

Jappie3 committed:
dovecot: 2.4.0 -> 2.4.2

Jappie3 committed:
dovecot_pigeonhole: add 2.4.2 alongside 0.5.21.1

Jappie3 committed:
nixos/dovecot: add settings option (rfc42)
